src/plugins/generic/validator.js

```diff
@@ -0,0 +1,23 @@
+var _ = require('lodash');
+var assert = require('assert');
+var util = require('util');
+
+function validateDatasource(datasource) {
+  assert(datasource.source !== undefined,
+    util.format('Missing source for generic datasource with id = %s', datasource.id));
+  assert(datasource.config !== undefined,
+    util.format('Missing config for generic datasource with id = %s', datasource.id));
+}
+
+function shouldHaveDatasourceId(widget) {
```

Comment:
This should do some more validation of the different widget types. If we add a folder for each widget, as suggested above, maybe we could add a validator.js for each folder and call that from here.